File: projAndrea/kalk/model/palette.cpp
#include "model/palette.h"

Palette::Palette(){
    size = 0;
}

Palette::Palette(const int& dim){
    size=dim;
    for(int i=0; i<size; i++){
        colori.push_back(Color());
    }
}

Palette::Palette(const int& dim, const Color & costr){
    size=dim;
    for(int i=0; i<size; i++){
        colori.push_back(costr);
    }
}

Palette::Palette(const Palette& copia){
    if(this != &copia){
        size=copia.size;
        colori.clear();
        auto it = copia.colori.begin();
        for(int i=0; i<size; i++){
            colori.push_back(it[i]);
        }
    }
}

//Palette::~Palette(){}

void Palette::opacity(const int & valore){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        it[i].setOpacity(static_cast<unsigned int>(valore));
    }
}

void Palette::add(const Color& addendo){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        it[i]=it[i]+addendo;
    }
}

void Palette::complementare(){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        it[i]=it[i].Complementare();
    }
}

void Palette::findReplace(const Color& cerca, const Color& sostituisci){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        if(it[i]==cerca){
            it[i]=sostituisci;
        }
    }
}

void Palette::desat(){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        it[i]=it[i].Desaturazione();
    }
}

void Palette::rimuovi(const int& index){
    --size;
    auto it = colori.begin();
    colori.erase (it+index);
}

void Palette::inserisci(const Color& colore){
    ++size;
    colori.push_back(Color (colore));
}

void Palette::fill(const Color& sostituisci){
    auto it = colori.begin();
    for (int i=0; i<size; i++){
        it[i]=sostituisci;
    }
}

void Palette::sostituisci(const Color& sostituisci, const int& posizione){
    auto it = colori.begin();
    it[posizione]=sostituisci;

}

int Palette::returnIndexMax(){
    auto it = colori.begin();
    int maxlocale = 0;
    for (int i=1; i<size; i++){
        if(it[i]>it[maxlocale]){
            maxlocale=i;
        }
    }
    return maxlocale;
}

int Palette::returnIndexMin(){
    auto it = colori.begin();
    int minlocale = 0;
    for (int i=1; i<size; i++){
        if(it[i]<it[minlocale]){
            minlocale=i;
        }
    }
    return minlocale;
}

int Palette::getRedFrom(const int& index){
    return static_cast<int>((colori.at(index)).getRed());
}

int Palette::getBlueFrom(const int& index){
    return static_cast<int>((colori.at(index)).getBlue());
}

int Palette::getGreenFrom(const int& index){
    return static_cast<int>((colori.at(index)).getGreen());
}

int Palette::getOpacityFrom(const int& index){
    return static_cast<int>((colori.at(index)).getOpacity());
}

void Palette::setRedAt(const int& index, const int& valore){
    colori.at(index).setRed(static_cast<unsigned int>(valore));
}

void Palette::setBlueAt(const int& index, const int& valore){
    colori.at(index).setBlue(static_cast<unsigned int>(valore));
}

void Palette::setGreenAt(const int& index, const int& valore){
    colori.at(index).setGreen(static_cast<unsigned int>(valore));
}

void Palette::setOpacityAt(const int& index, const int& valore){
    colori.at(index).setOpacity(static_cast<unsigned int>(valore));
}

Palette::~Palette(){

}


//aggiungere funzioni riordinaDecr e riordinaCr

File: projAndrea/Kalk Java/Palette.java
import java.util.Vector;

public class Palette implements ColorGroup {
	private int size;
	private Vector<Color> colori;
	
	public Palette(){
		size=0;
	}
	
	public Palette(int dim){
		size=dim;
		for(int i=0;i<dim;i++){
			colori.add(new Color());
		}
	}
	
	public Palette(int dim , Color c){
		size=dim;
		for(int i=0;i<size;i++){
			colori.add(c);
		}
	}
	
	public Palette(Palette c){
		size=c.size;
		colori=new Vector<Color>();
		for(int i=0;i<size;i++){
			colori.add(c.colori.get(i));
		}
	}
	
	public void opacity(int valore) {
		for(int i=0;i<size;i++){
			colori.get(i).setOpacity(valore);
		}
	}
	
	public void add(Color colore) {
		for(int i=0;i<size;i++){
			colori.get(i).add(colore);
		}
	}
	
	public void fill(Color colore) {
		for(int i=0;i<size;i++){
			colori.add(i, colore);
		}
	}
	
	public void sostituisci(int i , Color c){
		colori.add(i, c);
	}

	public void complementare() {
		for(int i=0;i<size;i++){
			colori.get(i).Complementare();
		}
	}

	public void findReplace(Color cerca , Color sostituisci) {
		for(int i=0;i<size;i++){
			if(colori.get(i).equals(cerca)){
				colori.add(i, sostituisci);
			}
		}
	}

	public void desat() {
		for(int i=0;i<size;i++){
			colori.get(i).Desaturazione();
		}
	}
	
	public void rimuovi(){
		size--;
		colori.remove(colori.size()-1);
	}
	
	public void inserisci(Color agg){
		size++;
		colori.add(agg);
	}
	
	public int returnIndexMax(){
		int maxLocale =0;
		for(int i=0;i<size;i++){
			if(colori.get(i).maggiore(colori.get(maxLocale))){
				maxLocale=i;
			}
		}
		return maxLocale;
	}
	
	public int returnIndexMin(){
		int minLocale =0;
		for(int i=0;i<size;i++){
			if(colori.get(i).minore(colori.get(minLocale))){
				minLocale=i;
			}
		}
		return minLocale;
	}
	
	public int getRedFrom(int i){
		return colori.get(i).getRed();
	}
	
	public int getGreenFrom(int i){
		return colori.get(i).getGreen();
	}
	
	public int getBlueFrom(int i){
		return colori.get(i).getBlue();
	}
	
	public int getOpacityFrom(int i){
		return colori.get(i).getOpacity();
	}
	
	public void setRedFrom(int i,int c){
		colori.get(i).setRed(c);
	}
	
	public void setGreenFrom(int i,int c){
		 colori.get(i).setGreen(c);
	}
	
	public void setBlueFrom(int i,int c){
		 colori.get(i).setBlue(c);
	}
	
	public void setOpacityFrom(int i,int c){
		 colori.get(i).setOpacity(c);
	}
	@Override
	public String toString(){
		String s="Colore: ";
		for(int i=0;i<size;i++){
			s=s+colori.get(i).toString();
			s=s+" Colore: ";
		}
		return s;
	}
}
